### DISCUSSION TITLE
*begin(stmt) fails with single row results

### DISCUSSION THREAD

**Discussion by xiptos** [geekyMrK/SQLiteCpp#2] [General]
I believe trying to get a row from a single row query result will result in:
```
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
```
when using `begin(stmt)` to retrieve the `rowIterator`.

Example:
```
    try {
        //Creates a SQLite connection object
        Sqlite::SqliteConnection connection("myProfile.db");

        // sqliteExecute(connectionOb, "query"); executes a SQLite query, in this case creates a table
        sqliteExecute(connection, "create table if not exists myResume (skills text, proficiency int DEFAULT 5)");

        sqliteExecute(connection, "insert into myResume(skills, proficiency ) values (?, ?)", "C++", 7);

        auto stmt = Sqlite::SqliteStatement(connection, "select skills, proficiency from myResume");
        stmt.execute();
        auto row = *begin(stmt);
        std::cout << typeid(row).name() << std::endl;
        std::cout << std::left << std::setw(8) << row.getString() << " : " << row.getInt(1) << std::endl;
    }
    catch (const Sqlite::exception &e) {
        std::cout << e.errorMessage_ << '(' << e.errorCode_ << ')';
    }
```

**Comment by geekyMrK** [comment]
The problem doesn't arise in `auto row = *begin(stmt);`, the problem is in next line, `cpp 
std::cout << typeid(row).name() << std::endl;`
As you are asking for the typeid of the row. which it currently didnt support yet.

### **there is a better way to do get a single row:**
```cpp
Sqlite::SqliteStatement stmt(connection, "select skills, proficiency from myResume"); 
stmt.execute();
std::cout << std::left << std::setw(8) << stmt.getString() << " : " << stmt.getInt(1) << std::endl;
//As the SqliteStatement class also has `get()` methods
```

Some implementation notes (if you are curious about the type of row object): 
1.) I use CRTP (Curiously Recurring Template Pattern which has zero runtime overhead) in which i have a base class SqliteReader. The SqliteReader has all the `get()` methods. The SqliteStatement and SqliteRow class inherit from the SqliteReader.
2.) The `begin(SqliteStatement)` method returns rowIterator class. rowIterator class just hold a pointer to SqliteStatement class. So, it's just a light-weight class to check whether the SqliteStatement has more results to show.
3.) The type you get by dereferencing the `begin(SqliteStatement)` is the rowIterator, but when you dereferencing it by using `*begin()`, you get a proxy type SqliteReader Class. 
4.) SqliteReader Class (as discussed in note 1) is inherited from SqliteReader Class. It is a lightweight class(uses CRTP) just contains a raw `sqlite3_stmt *` designed to iterate over the for loop (just has `get()` ) methods to retrieve the result.


**This is done to make it simple yet efficient (almost zero cost abstraction) C++ Wrapper over SQLite3** 🙂

**Reply by xiptos** [reply]
Got it! Thank you, @geekyMrK. Very nice approach! :)
